### PR TITLE
local-ai: 2.12.4 -> 2.13.0

### PR DESCRIPTION
**Pull request by ck3d**
## Description of changes

update to latest version and added more tests

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/lo/local-ai/package.nix [hunk 1]
```diff
@@ -371,18 +373,18 @@ let
...
 
   pname = "local-ai";
-  version = "2.12.4";
+  version = "2.13.0";
   src = fetchFromGitHub {
     owner = "go-skynet";
     repo = "LocalAI";
-    rev = "v${version}";
-    hash = "sha256-piu2B6u4ZfxiOd9SXrE7jiiiwL2SM8EqXo2s5qeKRl0=";
+    rev = "V${version}";
```
Comment (teto) [inline pkgs/by-name/lo/local-ai/package.nix:380]: should be `v`

Comment (teto) [reply pkgs/by-name/lo/local-ai/package.nix:380]: which makes me update the hash

Comment (ck3d) [reply pkgs/by-name/lo/local-ai/package.nix]: Thanks for testing. The project changed the tag name.
I updated the PR.

**Comment by teto** [comment]
when building I got an error about getGoDirs
```
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.4
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading github.com/josharian/intern v1.0.0
buildPhase completed in 2 minutes 9 seconds
Running phase: checkPhase
/nix/store/9wnvhjyxjykwn5y06xc9a2h8rs5fbfia-stdenv-linux/setup: line 1583: getGoDirs: command not found
Running phase: installPhase
```
and the tests timed out
```
&{PredictionOptions:{Model:Meta-Llama-3-8B-Instruct-Q4_K_M.gguf Language: N:0 TopP:0xc00025cd88 TopK:0xc00025cd80 Temperature:0xc00025cd60 Maxtokens:0xc00025cd58 Echo:false Batch:0 IgnoreEOS:false RepeatPenalty:0 Keep:0 FrequencyPenalty:0 PresencePenalty:0 TFZ:0xc00025ce30 TypicalP:0xc00025ce28 Seed:0xc00025ce50 NegativePrompt: RopeFreqBase:0 RopeFreqScale:0 NegativePromptScale:0 UseFastTokenizer:false ClipSkip:0 Tokenizer:} Name:gpt-3.5-turbo F16:0xc00025cdf0 Threads:0xc00025cde8 Debug:0xc000377290 Roles:map[] Embeddings:false Backend: TemplateConfig:{Chat:gpt-3.5-turbo_chat ChatMessage:gpt-3.5-turbo_chat_message Completion: Edit: Functions: UseTokenizerTemplate:false} PromptStrings:[A long time ago in a galaxy far, far away] InputStrings:[] InputToken:[] functionCallString: functionCallNameString: FunctionsConfig:{DisableNoAction:false NoActionFunctionName: NoActionDescriptionName: ParallelCalls:false NoGrammar:false ResponseRegex:} FeatureFlag:map[] LLMConfig:{SystemPrompt: TensorSplit: MainGPU: RMSNormEps:0 NGQA:0 PromptCachePath: PromptCacheAll:false PromptCacheRO:false MirostatETA:0xc00025ce20 MirostatTAU:0xc00025ce18 Mirostat:0xc00025ce10 NGPULayers:0xc00025ce40 MMap:0xc00025ce48 MMlock:0xc00025ce49 LowVRAM:0xc00025ce49 Grammar: StopWords:[<|eot_id|>] Cutstrings:[] TrimSpace:[] TrimSuffix:[] ContextSize:0xc00025cc70 NUMA:false LoraAdapter: LoraBase: LoraScale:0 NoMulMatQ:false DraftModel: NDraft:0 Quantization: GPUMemoryUtilization:0 TrustRemoteCode:false EnforceEager:false SwapSpace:0 MaxModelLen:0 TensorParallelSize:0 MMProj: RopeScaling: ModelType: YarnExtFactor:0 YarnAttnFactor:0 YarnBetaFast:0 YarnBetaSlow:0} AutoGPTQ:{ModelBaseName: Device: Triton:false UseFastTokenizer:false} Diffusers:{CUDA:false PipelineType: SchedulerType: EnableParameters: CFGScale:0 IMG2IMG:false ClipSkip:0 ClipModel: ClipSubFolder: ControlNet:} Step:0 GRPC:{Attempts:0 AttemptsSleepTime:0} VallE:{AudioPath:} CUDA:false DownloadFiles:[] Description: Usage:}
       > machine # [ 2521.609212] local-ai[641]: 12:24AM DBG Model already loaded in memory: Meta-Llama-3-8B-Instruct-Q4_K_M.gguf
       > machine # [ 2521.623600] local-ai[641]: 12:24AM DBG Model 'Meta-Llama-3-8B-Instruct-Q4_K_M.gguf' already loaded
       > machine # [ 2521.692898] local-ai[641]: 12:24AM DBG GRPC(Meta-Llama-3-8B-Instruct-Q4_K_M.gguf-127.0.0.1:44219): stdout {"timestamp":1714350278,"level":"INFO","function":"launch_slot_with_data","line":884,"message":"slot is processing task","slot_id":0,"task_id":57}
       > machine # [ 2521.695825] local-ai[641]: 12:24AM DBG GRPC(Meta-Llama-3-8B-Instruct-Q4_K_M.gguf-127.0.0.1:44219): stdout {"timestamp":1714350278,"level":"INFO","function":"update_slots","line":1783,"message":"kv cache rm [p0, end)","slot_id":0,"task_id":57,"p0":0}
       > timeout reached; test terminating...
       > kill machine (pid 9)
       > qemu-kvm: terminating on signal 15 from pid 6 (/nix/store/glfr70gi7hfaj50mwj2431p8bg60fhqw-python3-3.11.9/bin/python3.11)
       > cleanup
```
Tests are nice when they work but I dont think it's a reason to block the MR (else you punish people writing tests). I've tested local-ai myself and it worked. I tried to enable the service.local-ai but it's not public yet I think ? (just used in tests ?). It failed eval. 
Looks like it's the first release with the CLI parser changed for kong following one of my request for a better local-ai configuration so I am happy :)

**Review by teto — APPROVED** [review]

**Comment by ck3d** [comment]
I am a little bit surprised, all the tests run on my old machine. The llama test takes 1:22 minutes.
Thanks for testing. I will merge this PR.

**Comment by teto** [comment]
I enabled local-ai locally via
```
  imports = [
    "${modulesPath}/../../pkgs/by-name/lo/local-ai/module.nix"
  ];

  services.local-ai.enable = true;
```
is this a conscious decision not to expose the module for now ? dont want to deal with the process or are you waiting for it to be better ? It could benefit from a freeformat `settings` now that localai config can live in a file.

**Comment by ck3d** [comment]
I started to write the module to simplify the NixOS tests. I decided to have declarative modules, no state directory, and dynamic user.
I am not sure if this is a good decision for a NixOS module. When we move the NixOS module to nixos/, then the tests should move to nixos/tests/. I avoided it for now, to keep fast updates without backward compatibility issues.
We can face the mode after 24.05 was released.

**Pull request merged**